src/fluree/db/json_ld/credential.cljc

```diff
@@ -8,180 +8,176 @@
 
-
 (defn generate
-  [credentialSubject opts]
+  "Generate a VerifiableCredential given a subject and some issuer opts."
+  [credential-subject opts]
   (let [{:keys [did private]} opts
         did* (or (:id did)
                  (str "did:fluree:" (crypto/account-id-from-private private)))]
-    (sign
-      {"@context"          ["https://www.w3.org/2018/credentials/v1"
+    (add-proof {"@context" ["https://www.w3.org/2018/credentials/v1"
                             "https://flur.ee/ns/block"]
-       "id"                "blah"
-       "type"              ["VerifiableCredential"]
-       "issuer"            did*
-       "issuanceDate"      (util/current-time-iso)
-       "credentialSubject" credentialSubject}
-      private)))
+                ;; TODO: use a useful id
+                "id" "blah"
+                "type" ["VerifiableCredential"]
+                "issuer" did*
+                "issuanceDate" (util/current-time-iso)
+                "credentialSubject" credential-subject}
 
+               private)))
 
-(defn verify
+(defn decode-credential
   [credential-json]
-  (let [cred            (json/parse credential-json false)
-        cred-normalized (-> cred
-                            (dissoc "proof")
-                            json-ld/normalize-data)
-        signing-input   (signing-input cred-normalized)
-        proof           (get cred "proof")
-        proof-did       (get proof "verificationMethod")
-        jws             (get proof "jws")                   ;; "eyJhbGciOiJFUzI1NkstUiIsImI2NCI6ZmFsc2UsImNyaXQiOlsiYjY0Il19..<signature data>"
-        [header signature] (str/split jws #"\.\.")
-        sig-hex         (alphabase/base-to-base signature :base64 :hex)
-        key-id          (crypto/account-id-from-message signing-input sig-hex)
-        derived-did     (str "did:fluree:" key-id)]
-    (when-not (= jws-header-b64 header)
-      ;; TODO - below should use :base64URL once supported
-      (throw (ex-info (str "Unsupported jws header in credential: " (alphabase/base-to-base header :base64 :string))
-                      {:status 403 :error :json-ld/invalid-credential})))
-    (when-not (= derived-did proof-did)
-      (throw (ex-info (str "Derived did from signature does not match did in 'proof' of credential. Derived: "
-                           derived-did ", proof verificationMethod: " proof-did)
-                      {:status 403 :error :json-ld/invalid-credential})))
-    cred))
-
+  (json/parse credential-json false))
 
+(defn verify
+  "Takes a credential"
+  [credential]
+  (let [proof-did     (get-in credential ["proof" "verificationMethod"])
+        jws           (get-in credential ["proof" "jws"])
+        {:keys [header payload signature]} (deserialize-jws jws)
+        signing-input (jws-signing-input payload)
+        key-id        (crypto/account-id-from-message signing-input signature)
+        derived-did   (str "did:fluree:" key-id)]
+    (cond-> {:credential credential}
```

Comment:
I changed the validation to not throw exceptions, since I tend to prefer to handle non-exceptional circumstances with just data, though if exceptions mesh better with the overall flow I'm happy to go back to them. In this case, you just check if there are any `:errors` in the return map.